tfprotov5/internal/tf5serverlogging/downstream_request.go

```diff
@@ -0,0 +1,40 @@
+package tf5serverlogging
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-go/internal/logging"
+	"github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/diag"
+)
+
+// DownstreamRequest sets a request duration start time context key and
+// generates a TRACE "Sending request downstream" log.
+func DownstreamRequest(ctx context.Context) context.Context {
+	requestStart := time.Now()
+	ctx = context.WithValue(ctx, ContextKeyDownstreamRequestStartTime{}, requestStart)
```

Comment:
It wouldn't throw an error, but rather prevents unintentional context value updates, by being proactive about a context key issue upfront.

Using a `string` type means that anyone can overwrite your value, if they know the `string` value for the key. e.g.

```go
ctx = context.WithValue(ctx, "sweet", "first")

// later on... past module boundaries, etc.

ctx = context.WithValue(ctx, "sweet", "updated")
```

Folks calling `ctx.Value("sweet")` would then get the value "updated" without knowing the original value was modified. Imagine now that folks finding a key like "request-start-time" could then overwrite the SDK defined request start time, for whatever reason (it would also require the SDK to get the returned context after provider logic, which we generally wouldn't want to do today, but maybe there is a use case for it in the future where this could bite us). 

Using a new type as the context key, generally a `struct{}` to prevent actual memory allocation, is scoped to the package and type. It's therefore harder to have other Go modules or packages modify context values unintentionally or in the case of `internal/` scoped packages, not even possible.

Reply:
Makes perfect sense. Thank you.